hash/blake2b.go

```diff
@@ -0,0 +1,78 @@
+// Go Substrate RPC Client (GSRPC) provides APIs and types around Polkadot and any Substrate-based chain RPC calls
+//
+// Copyright 2019 Centrifuge GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hash
+
+import (
+	"hash"
+
+	"golang.org/x/crypto/blake2b"
+)
+
+type blake2b128Concat struct {
+	hasher hash.Hash
+	data   []byte
+}
+
+// NewBlake2b128Concat returns an instance of blake2b concat hasher
+func NewBlake2b128Concat(k []byte) (hash.Hash, error) {
+	h, err := blake2b.New(16, k)
+	if err != nil {
+		return nil, err
+	}
+	return &blake2b128Concat{hasher: h, data: k}, nil
+}
+
+// Write (via the embedded io.Writer interface) adds more data to the running hash.
+func (bc *blake2b128Concat) Write(p []byte) (n int, err error) {
+	bc.data = append(bc.data, p...)
+	return bc.hasher.Write(p)
+}
+
+// Sum appends the current hash to b and returns the resulting slice.
+// It does not change the underlying hash state.
+func (bc *blake2b128Concat) Sum(b []byte) []byte {
+	return append(bc.hasher.Sum(b), bc.data...)
```

Comment:
This looks weird. Do we want to append the `data` to the resultant hash?

Reply:
https://github.com/paritytech/substrate/blob/a063383fed22eb08703475317b742c84a8dc4cc3/frame/support/src/hash.rs#L116

Reply:
cool!